src/session_impl.cpp

```diff
@@ -811,35 +808,29 @@ namespace aux {
 
 		boost::shared_ptr<plugin> p(new session_plugin_wrapper(ext));
 
-		m_ses_extensions.push_back(p);
-		m_session_extension_features |= p->implemented_features();
+		add_ses_extension(p, false);
 	}
 
-	void session_impl::add_ses_extension(boost::shared_ptr<plugin> ext)
+	void session_impl::add_ses_extension(boost::shared_ptr<plugin> ext, bool no_wrap)
 	{
 		TORRENT_ASSERT(is_single_thread());
 		TORRENT_ASSERT_VAL(ext, ext);
 
-		m_ses_extensions.push_back(ext);
-		m_alerts.add_extension(ext);
-		ext->added(session_handle(this));
-		m_session_extension_features |= ext->implemented_features();
+		boost::uint32_t features = ext->implemented_features();
 
-		// get any DHT queries the plugin would like to handle
-		// and record them in m_extension_dht_queries for lookup
-		// later
-		dht_extensions_t dht_ext;
-		ext->register_dht_extensions(dht_ext);
-		for (dht_extensions_t::iterator e = dht_ext.begin();
-			e != dht_ext.end(); ++e)
+		m_ses_extensions[plugins_all_idx].push_back(ext);
+		if (features & plugin::optimistic_unchoke_feature)
+			m_ses_extensions[plugins_optimistic_unchoke_idx].push_back(ext);
+		if (features & plugin::tick_feature)
+			m_ses_extensions[plugins_tick_idx].push_back(ext);
+		if (features & plugin::dht_request_feature)
+			m_ses_extensions[plugins_dht_request_idx].push_back(ext);
+
+		if (no_wrap)
 		{
-			TORRENT_ASSERT(e->first.size() <= max_dht_query_length);
-			if (e->first.size() > max_dht_query_length) continue;
-			extension_dht_query registration;
-			registration.query_len = uint8_t(e->first.size());
-			std::copy(e->first.begin(), e->first.end(), registration.query.begin());
-			registration.handler = e->second;
-			m_extension_dht_queries.push_back(registration);
+			if (features & plugin::alert_feature)
+				m_alerts.add_extension(ext);
```

Comment:
I will add the comment of `no_wrap`, but I'm pretty sure is the name that it's wrong. Now, about the special nature of `alert_feature`, the reasoning here is that if you are a "wrap", you are 100% sure the plugin is not subscribed to alerts. And this is the same reason the call `added()` is inside this `if` too. Take a look at `session_impl::add_extension(ext_function_t ext)`:
```
boost::shared_ptr<plugin> p(new session_plugin_wrapper(ext));
add_ses_extension(p, false);
```